packages/dds/merge-tree/Obliterate.md

```diff
@@ -0,0 +1,276 @@
+
+This is easier said than done: concretely, and continuing with the example number given above, suppose this insertion happens:
+```
+{ seq: 60, refSeq: 40, clientId: 3, op: <insert "hello" at index 10> }
+```
+
+After locating the insertion point and updating the merge tree, we need to decide if the resulting segment is inside of an obliterated region.
+If we happened to know the `seq` of the obliteration we were testing for, this is easy: the first adjacent segment in each direction from
+the perspective of `{ seq: 50, clientId: localClientId }` can inform us if we're either inside or directly adjacent to an obliterated range.
+Thus, a naive implementation could check all sequence numbers in the collab window.
+The obvious optimization of only checking seq numbers of obliterate ops would improve this slightly.
+We can do asymptotically better by leveraging the tree structure.
+It would be ideal if we only needed to perform one commonly short excursion in each direction.
+The only candidate that makes much sense is from the perspective of `{ seq: 60, clientId: 3 }` (i.e. the client submitting the insert op at
+the time the op is sequenced).
+The problem with this perspective is that ops 51 through 59 may have inserted a segment between the inserted "hello"
+and the obliterated range that was submitted by a client which has already acked the obliterate.
+For example, `{ seq: 55, clientId: 5, refSeq: 50, op: <insert "i won't be obliterated" at index 10> }`.
+
+The forward excursion would need to continue past this segment in order to conclude it isn't in an obliterated range.
+If that was the only such concurrent insert, the next segment it would visit would be an obliterated one and we'd decide
+whether or not to include the newly inserted segment as part of the obliterated region based on some endpoint merge strategy.
+
+The key insight is that visiting the segment with seq 55 does provide the excursion with information: since the segment
+was inserted at seq 55 and isn't obliterated, any obliterate operation must have occurred before seq 55.
+If we keep track of the smallest sequence number that we've visited, we can halt the excursion as soon as it falls below
+the smallest obliterate operation within the collab window.
+If we alternatively reach a segment that has been obliterated concurrently to the insert we're processing, we can also stop
+and use the endpoint resolution strategy.
+
+All-in-all, the insert logic modification might look something like this:
+
+```typescript
+function insertingWalk(args /* mostly omitted */, op) {
+  /* regular insert logic goes here */
+
+  let currentMin = Number.POSITIVE_INFINITY;
+  let obliteratedSegment: ISegment | undefined = undefined;
+  const smallestSeqObliterateOp = this.getSmallestSeqObliterateOp();
+  const findAdjacentObliteratedSegment = (seg) => {
+    if (seg.seq === UnassignedSequenceNumber) {
+      // Ignore un-acked segments
+      return true;
```

Comment:
i need to think more, but i don't think this can happen in practice

Reply:
Hmm, yeah this might be right. I had originally thought this was possible when walking backward in response to a remote insert if there was a local insert in the middle, but I don't think it's ever possible for that to happen. The minimum check needs to be adjusted to remove this clause though.

Reply:
Removed it for now. We can let fuzz tell us if we're right when we go to implement ;)